### PR TITLE
Fix binding logic for dictionaries with complex elements

### LINKED ISSUE #89010: Configuration binding source generator fails with CS8598 error
### Description

Attempting to enable the Configuration binding source generator with a struct used as a custom configuration type causes the source generator to emit code that fails to compile with a CS8598 error.

```
Repro\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\GeneratedConfigurationBinder.g.cs(122,17): error CS8598: The suppression operator is not allowed in this context [Repro\Repro.csproj]
```

The method generated is below. The failing line is `element! = temp4;`.

```csharp
public static void BindCore(IConfiguration configuration, ref IDictionary<string, Geolocation> obj, BinderOptions? binderOptions)
{
    if (obj is null)
    {
        throw new ArgumentNullException(nameof(obj));
    }

    foreach (IConfigurationSection section in configuration.GetChildren())
    {
        if (!(obj.TryGetValue(section.Key!, out Geolocation element)))
        {
            element = InitializeGeolocation(section, binderOptions);;
        }
        var temp4 = InitializeGeolocation(configuration, binderOptions);;
        BindCore(section, ref temp4, binderOptions);
        element! = temp4; // CS8598 occurs here
        obj[section.Key!] = element;
    }
}
```

### Reproduction Steps

1. Clone [martincostello/config-binding-source-generator-cs8598-repro](https://github.com/martincostello/config-binding-source-generator-cs8598-repro)
1. Run `dotnet build`

### Expected behavior

The application compiles and configuration binding works equivalently to runtime binding.

### Actual behavior

The application fails to compile with a CS8598 error.

### Regression?

No.

### Known Workarounds

None.

### Configuration

.NET SDK 8.0.100-preview.6.23330.14

### Other information

_No response_

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/runtime/issues/89010.

FYI @martincostello.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs [hunk 1]
```diff
@@ -1900,5 +1900,70 @@ public void AllowsCaseInsensitiveMatch()
...
+        }
+
+        [Fact]
+        public void ComplexObj_As_Dictionary_Element()
+        {
+            var configuration = TestHelpers.GetConfigurationFromJsonString("""
+                {
+                    "First":
+                    {
+                        "Latitude": 3,
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs:1934]: Is it worth try a configuration providing only Latitude and not necessary the Longitude?

**Review by tarekgh — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs:1934]: We have a few tests for missing configs + ctor params - https://github.com/dotnet/runtime/blob/dfe7a6c9749ce3c9be9bf99f544fb13c04c61742/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs#L958-L1018.

**Comment by ViktorHofer** [comment]
@layomia this broke the build. The `runtime (Build Libraries Build windows net48 x86 Release)` leg was red. I will revert the change.

**Comment by layomia** [comment]
@ViktorHofer oops. Thanks.

**Pull request merged**